File: src/services/customCommand.ts
import { invoke } from '@tauri-apps/api/core';
import type { CustomCommand } from '@/types';

export interface CustomCommandInput {
  id?: string;
  name: string;
  command: string;
  description?: string;
  color?: string;
  createdAt?: string;
}

export async function saveCustomCommand(config: CustomCommandInput): Promise<CustomCommand> {
  const now = new Date().toISOString();

  const payload: Record<string, string> = {
    name: config.name,
    command: config.command,
    description: config.description || '',
    color: config.color || '',
  };
  if (config.id) {
    payload.id = config.id;
  }

  const result = await invoke<{ id: string }>('save_custom_command', { config: payload });
  return {
    id: result.id,
    name: config.name,
    command: config.command,
    description: config.description,
    color: config.color || '',
    createdAt: config.createdAt || now,
    updatedAt: now,
  };
}

export async function getCustomCommands(): Promise<CustomCommand[]> {
  return invoke<CustomCommand[]>('get_custom_commands');
}

export async function deleteCustomCommand(id: string): Promise<void> {
  await invoke('delete_custom_command', { id });
}

File: src-tauri/src/custom_command.rs
use chrono::Local;
use uuid::Uuid;

use crate::db::connection::DbManager;
use crate::db::repo;
use crate::models::types::{AppError, CustomCommand};

/// Save a custom command to the database.
/// If the command has an existing ID, it will be updated.
/// Otherwise, a new ID is generated.
pub fn save_custom_command(
    db: &DbManager,
    config: serde_json::Value,
) -> Result<String, AppError> {
    let conn = db.lock_conn()?;
    let now = Local::now().to_rfc3339();

    let id: String = config
        .get("id")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string())
        .unwrap_or_else(|| Uuid::new_v4().to_string());

    let name: String = config
        .get("name")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string())
        .unwrap_or_default();

    let command: String = config
        .get("command")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string())
        .unwrap_or_default();

    let description: Option<String> = config
        .get("description")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string());

    let shell_type: String = config
        .get("shellType")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string())
        .unwrap_or_else(|| "cmd".to_string());

    let custom_command = CustomCommand {
        id: id.clone(),
        name,
        command,
        description,
        shell_type,
        created_at: now.clone(),
        updated_at: now,
    };

    if repo::custom_command_exists(&conn, &id)? {
        repo::update_custom_command(&conn, &custom_command)?;
    } else {
        repo::insert_custom_command(&conn, &custom_command)?;
    }

    Ok(id)
}

/// Get all custom commands from the database.
pub fn get_custom_commands(
    db: &DbManager,
) -> Result<Vec<serde_json::Value>, AppError> {
    let conn = db.lock_conn()?;
    let commands = repo::get_all_custom_commands(&conn)?;

    let json_commands: Vec<serde_json::Value> = commands.into_iter().map(|c| {
        serde_json::json!({
            "id": c.id,
            "name": c.name,
            "command": c.command,
            "description": c.description,
            "shellType": c.shell_type,
            "createdAt": c.created_at,
            "updatedAt": c.updated_at,
        })
    }).collect();

    Ok(json_commands)
}

/// Delete a custom command by ID.
pub fn delete_custom_command(db: &DbManager, id: &str) -> Result<bool, AppError> {
    let conn = db.lock_conn()?;
    repo::delete_custom_command_by_id(&conn, id)
}
